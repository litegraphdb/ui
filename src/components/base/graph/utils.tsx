import { NodeData } from '@/lib/graph/types';
import { defaultNodeColor, nodeLightColorByType, universalNodeColors } from './constant';

const maxLegends = 16;

export const getLegendsForNodes = (nodes: NodeData[]) => {
  const priorityLegends = new Map<string, { legend: string; color: string }>();
  const customLegends = new Map<string, { legend: string; color: string }>();

  // Separate legends into priority (nodeLightColorByType) and custom
  nodes.forEach((node) => {
    if (Object.keys(nodeLightColorByType).includes(node.type)) {
      priorityLegends.set(node.type, { legend: node.type, color: nodeLightColorByType[node.type] });
    } else {
      // Only add if not already present to avoid changing colors for same legend
      if (!customLegends.has(node.type)) {
        customLegends.set(node.type, {
          legend: node.type,
          color: universalNodeColors[customLegends.size % universalNodeColors.length],
        });
      }
    }
  });

  const priorityLegendsSize = priorityLegends.size;
  // If there are nodes with unknown types, add "Unknown" to priority legends
  const isMoreThanMaxLegends = customLegends.size + priorityLegendsSize > maxLegends;

  // Convert to arrays for easier manipulation
  const priorityArray = Array.from(priorityLegends.values());
  const customArray = Array.from(customLegends.values()).slice(
    0,
    maxLegends - (priorityArray.length + (isMoreThanMaxLegends ? 1 : 0))
  );

  // Combine priority legends first, then custom legends
  const legends = [
    ...priorityArray,
    ...customArray,
    ...(isMoreThanMaxLegends ? [{ legend: 'Unknown', color: defaultNodeColor }] : []),
  ];
  const mapLegends: Record<string, { legend: string; color: string }> = {};
  legends.forEach((legend) => {
    mapLegends[legend.legend] = legend;
  });
  return mapLegends;
};
