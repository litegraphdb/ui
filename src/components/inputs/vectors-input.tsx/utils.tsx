export const convertVectorsToAPIRecord = (
  vectors?: Array<{ Model: string; Dimensionality: number; Content: string; Vectors: number[] }>
) => {
  return (
    vectors?.map((vector: any) => ({
      ...vector,
      Dimensionality: Number(vector.Dimensionality),
      Vectors: vector.Vectors || [],
    })) || []
  );
};
