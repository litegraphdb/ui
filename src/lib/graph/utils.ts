import type { NodeData, Point } from './types';

export function isPointInNode(point: Point, node: NodeData, radius: number = 10): boolean {
  const dx = point.x - node.x;
  const dy = point.y - node.y;
  return dx * dx + dy * dy <= radius * radius;
}

export function getNodeAtPosition(nodes: NodeData[], point: Point): NodeData | null {
  return nodes.find((node) => isPointInNode(point, node)) || null;
}

export function transformToOptions<T extends { GUID: string; name?: string; Name?: string }>(
  data?: T[] | null,
  labelField: keyof T = 'name' // Field to use for label in options
) {
  return data
    ? data?.map((item) => ({
        value: item.GUID,
        label: (item[labelField] as string) || item.Name || item.GUID,
      }))
    : [];
}
