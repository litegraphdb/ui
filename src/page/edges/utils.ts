import { EdgeType } from '@/lib/store/edge/types';
import { NodeType } from '@/lib/store/node/types';

// Utility to find the name by GUID
export const getNodeNameByGUID = (guid: string, nodesList: NodeType[]) => {
  if (!guid) return 'N/A';
  const node = nodesList.find((n: NodeType) => n.GUID === guid);
  return node ? node.Name : guid;
};

export const transformEdgeDataForTable = (edgesList: EdgeType[], nodesList: NodeType[]) => {
  return (
    edgesList?.map((data: EdgeType) => {
      return {
        ...data,
        FromName: getNodeNameByGUID(data.From, nodesList),
        ToName: getNodeNameByGUID(data.To, nodesList),
        key: data.GUID,
      } as EdgeType;
    }) || []
  );
};
