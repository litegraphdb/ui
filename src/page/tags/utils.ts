import { NodeType } from '@/lib/store/node/types';

import { getNodeNameByGUID } from '../edges/utils';
import { EdgeType } from '@/lib/store/edge/types';
import { TagMetaData } from 'litegraphdb/dist/types/types';
import { TagType } from '@/lib/store/tag/types';

// Utility to find the name by GUID
export const getEdgeNameByGUID = (guid: string, edgesList: EdgeType[]) => {
  if (!guid) return 'N/A';
  const edge = edgesList.find((n: EdgeType) => n.GUID === guid);
  return edge ? edge.Name : guid;
};

export const transformTagsDataForTable = (
  tagsList: TagMetaData[],
  nodesList: NodeType[],
  edgesList: EdgeType[]
) => {
  return (
    tagsList?.map((data: TagMetaData) => {
      return {
        ...data,
        NodeName: getNodeNameByGUID(data.NodeGUID, nodesList),
        EdgeName: getEdgeNameByGUID(data.EdgeGUID, edgesList),
        key: data.GUID,
      } as TagType;
    }) || []
  );
};
