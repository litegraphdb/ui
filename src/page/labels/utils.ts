import { NodeType } from '@/types/types';
import { getNodeNameByGUID } from '../edges/utils';

import { getEdgeNameByGUID } from '../tags/utils';
import { LabelMetadata } from 'litegraphdb/dist/types/types';
import { EdgeType } from '@/types/types';
import { LabelMetadataForTable } from './types';

export const transformLabelsDataForTable = (
  labelsList: LabelMetadata[],
  nodesList: NodeType[],
  edgesList: EdgeType[]
) => {
  return (
    labelsList?.map((data: LabelMetadata) => {
      return {
        ...data,
        NodeName: getNodeNameByGUID(data.NodeGUID, nodesList),
        EdgeName: getEdgeNameByGUID(data.EdgeGUID, edgesList),
        key: data.GUID,
      } as LabelMetadataForTable;
    }) || []
  );
};

export const getNodeAndEdgeGUIDsByLabelList = (labelsList: LabelMetadata[]) => {
  const nodeGUIDs = labelsList.map((label) => label.NodeGUID);
  const edgeGUIDs = labelsList.map((label) => label.EdgeGUID);
  return {
    nodeGUIDs: Array.from(new Set(nodeGUIDs)),
    edgeGUIDs: Array.from(new Set(edgeGUIDs)),
  };
};
