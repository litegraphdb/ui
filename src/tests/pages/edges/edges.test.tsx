import { screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import { renderWithRedux } from '../../store/utils';
import { createMockStore } from '../../store/mockStore';
import { fireEvent } from '@testing-library/react';
import { mockGraphData } from '../mockData';
import { LiteGraphSdk } from 'litegraphdb';
import EdgePage from '@/app/dashboard/[tenantId]/edges/page';
import userEvent from '@testing-library/user-event';
import { setupServer } from 'msw/node';
import { handlers } from './handler';
import { commonHandlers } from '@/tests/handler';
import { setTenant } from '@/lib/sdk/litegraph.service';
import { mockTenantGUID } from '../mockData';

const server = setupServer(...handlers, ...commonHandlers);
setTenant(mockTenantGUID);

let container: any;
describe.skip('EdgePage with Mock API', () => {
  beforeEach(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());
  const store = createMockStore();

  it('renders the edge page title', async () => {
    const wrapper = renderWithRedux(
      <EdgePage />,
      {
        ...store,
        graphsList: {
          graphs: mockGraphData,
        },
      } as any,
      true
    );

    // Verify the Nodes heading
    const heading = screen.getByTestId('heading');
    expect(heading).toHaveTextContent('Edges');

    // Verify the select dropdown for graphs
    expect(wrapper.container).toMatchSnapshot();
    const graphSelect = screen.getByTestId('litegraph-select');
    expect(graphSelect).toBeInTheDocument();
    expect(graphSelect).toHaveTextContent('Test Demo Graphtestttt 2');

    waitFor(() => {
      // Simulate graph selection
      fireEvent.change(graphSelect, { target: { value: 'd52aeab4-4de7-4076-98dd-461d4a61ac88' } });

      // Verify API Call
      expect((LiteGraphSdk as any).mock.instances[0].readEdges).toHaveBeenCalled();

      expect(screen.getByText('new test 23')).toBeInTheDocument();

      // Verify the create node button
      const createButton = screen.getByRole('button', { name: /create node/i });
      expect(createButton).toBeInTheDocument();
    });
    expect(wrapper.container).toMatchSnapshot();
  });

  it('renders the EdgePage and create edge', async () => {
    const user = userEvent.setup();

    renderWithRedux(
      <EdgePage />,
      {
        ...store,
        graphsList: {
          graphs: mockGraphData,
        },
      } as any,
      true
    );

    // Verify the Nodes heading
    const heading = screen.getByTestId('heading');
    expect(heading).toHaveTextContent('Edges');

    // Verify the select dropdown for graphs
    const graphSelect = screen.getByTestId('litegraph-select');
    expect(graphSelect).toBeInTheDocument();
    expect(graphSelect).toHaveTextContent('Test Demo Graphtestttt 2'); // Adjusted expectation for the mock

    waitFor(async () => {
      // Simulate graph selection
      fireEvent.change(graphSelect, { target: { value: 'd52aeab4-4de7-4076-98dd-461d4a61ac88' } });

      // Verify API Call
      expect((LiteGraphSdk as any).mock.instances[0].readEdges).toHaveBeenCalled();

      expect(screen.getByText('new test 23')).toBeInTheDocument();

      // Verify the create node button
      const createButton = screen.getAllByRole('button', { name: 'Create' });
      await user.click(createButton[0]);

      await user.click(createButton);

      // Fill in the Create Graph Form
      const nameInput = screen.getByTestId('edge-name-input');
      const dataEditor = screen.getByTestId('edge-data-input');

      await user.type(nameInput, 'New Edge');
      fireEvent.change(dataEditor, { target: { value: '{"edge":{}}' } });

      // Submit the Form
      const createNodeButton = screen.getByText('Create');
      await user.click(createNodeButton);

      // Wait for the new graph to appear
      expect(screen.getByText('New Edge')).toBeInTheDocument();

      // Verify API Call
      expect((LiteGraphSdk as any).mock.instances[0].createEdge).toHaveBeenCalledWith({
        GUID: expect.any(String),
        Name: 'New Edge',
        Data: { edge: {} },
      });
    });
  });

  it('renders the EdgePage and update edge', async () => {
    const user = userEvent.setup();

    renderWithRedux(
      <EdgePage />,
      {
        ...store,
        graphsList: {
          graphs: mockGraphData,
        },
      } as any,
      true
    );

    // Verify the Nodes heading
    const heading = screen.getByTestId('heading');
    expect(heading).toBeInTheDocument();

    // Verify the select dropdown for graphs
    const graphSelect = screen.getByTestId('litegraph-select');
    expect(graphSelect).toBeInTheDocument();
    expect(graphSelect).toHaveTextContent('Test Demo Graphtestttt 2'); // Adjusted expectation for the mockk

    waitFor(async () => {
      // Simulate graph selection
      fireEvent.change(graphSelect, { target: { value: 'd52aeab4-4de7-4076-98dd-461d4a61ac88' } });

      // Verify API Call
      expect((LiteGraphSdk as any).mock.instances[0].readEdges).toHaveBeenCalled();

      expect(screen.getByText('new test 23')).toBeInTheDocument();

      // Verify the update node button
      const updateButton = screen.getAllByRole('button', { name: 'Edit' });
      await user.click(updateButton[0]);

      await user.click(updateButton);

      // Fill in the Create Graph Form
      const nameInput = screen.getByTestId('node-name-input');
      const dataEditor = screen.getByTestId('node-data-input');

      await user.type(nameInput, 'New Updated Node');
      fireEvent.change(dataEditor, { target: { value: '{"node":{}}' } });

      // Submit the Form
      const UpdateNodeButton = screen.getByText('Update');
      await user.click(UpdateNodeButton);

      // Wait for the new graph to appear
      expect(screen.getByText('New Updated Node')).toBeInTheDocument();

      // Verify API Call
      expect((LiteGraphSdk as any).mock.instances[0].createNode).toHaveBeenCalledWith({
        GUID: expect.any(String),
        Name: 'New Updated Node',
        Data: { graph: {} },
      });
    });
  });

  it('render fallback message on edge load error', async () => {
    const wrapper = renderWithRedux(
      <EdgePage />,
      {
        ...store,
      } as any,
      true
    );

    waitFor(() => {
      expect(screen.getByText('Something went wrong.')).toBeInTheDocument();
    });

    expect(wrapper.container).toMatchSnapshot();
  });
});
