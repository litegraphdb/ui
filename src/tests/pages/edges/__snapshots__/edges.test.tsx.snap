// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EdgePage with Mock API creates a new edge: final form state 1`] = `
<div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;

exports[`EdgePage with Mock API creates a new edge: form filled with test data 1`] = `
<div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;

exports[`EdgePage with Mock API creates a new edge: initial modal state 1`] = `
<div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;

exports[`EdgePage with Mock API renders fallback on error: fallback message 1`] = `
<div>
  <main
    class="ant-layout-content pageContainer css-dev-only-do-not-override-98ntnt"
    id="edges"
  >
    <div
      class="pageTitleContainer ant-flex css-dev-only-do-not-override-98ntnt ant-flex-wrap-wrap ant-flex-align-center ant-flex-justify-space-between ant-flex-gap-small"
    >
      <span
        class="ant-typography css-dev-only-do-not-override-98ntnt"
        data-testid="heading"
        style="font-weight: 600; font-size: 16px;"
      >
        <div
          class="ant-flex css-dev-only-do-not-override-98ntnt ant-flex-align-center"
          style="gap: 10px;"
        >
          <span
            class="ant-typography css-dev-only-do-not-override-98ntnt"
          >
            Edges
          </span>
          <span
            aria-label="search"
            class="anticon anticon-search cursor-pointer"
            role="img"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              data-icon="search"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
              />
            </svg>
          </span>
        </div>
      </span>
      <button
        class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-link"
        style="font-weight: 500;"
        type="button"
      >
        <span
          class="ant-btn-icon"
        >
          <span
            aria-label="plus-square"
            class="anticon anticon-plus-square"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="plus-square"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
              />
              <path
                d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
              />
            </svg>
          </span>
        </span>
        <span>
          Create Edge
        </span>
      </button>
    </div>
    <div
      class="pageContent"
    >
      <div
        class="ant-flex css-dev-only-do-not-override-98ntnt ant-flex-align-center ant-flex-justify-center ant-flex-vertical"
      >
        <span
          class="ant-typography css-dev-only-do-not-override-98ntnt"
        >
          Something went wrong.
        </span>
        <span
          aria-label="close-circle"
          class="anticon anticon-close-circle colorRed icon"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="close-circle"
            fill="currentColor"
            fill-rule="evenodd"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64c247.4 0 448 200.6 448 448S759.4 960 512 960 64 759.4 64 512 264.6 64 512 64zm0 76c-205.4 0-372 166.6-372 372s166.6 372 372 372 372-166.6 372-372-166.6-372-372-372zm128.01 198.83c.03 0 .05.01.09.06l45.02 45.01a.2.2 0 01.05.09.12.12 0 010 .07c0 .02-.01.04-.05.08L557.25 512l127.87 127.86a.27.27 0 01.05.06v.02a.12.12 0 010 .07c0 .03-.01.05-.05.09l-45.02 45.02a.2.2 0 01-.09.05.12.12 0 01-.07 0c-.02 0-.04-.01-.08-.05L512 557.25 384.14 685.12c-.04.04-.06.05-.08.05a.12.12 0 01-.07 0c-.03 0-.05-.01-.09-.05l-45.02-45.02a.2.2 0 01-.05-.09.12.12 0 010-.07c0-.02.01-.04.06-.08L466.75 512 338.88 384.14a.27.27 0 01-.05-.06l-.01-.02a.12.12 0 010-.07c0-.03.01-.05.05-.09l45.02-45.02a.2.2 0 01.09-.05.12.12 0 01.07 0c.02 0 .04.01.08.06L512 466.75l127.86-127.86c.04-.05.06-.06.08-.06a.12.12 0 01.07 0z"
            />
          </svg>
        </span>
        <span
          class="ant-typography text-link css-dev-only-do-not-override-98ntnt"
        >
          Retry
        </span>
      </div>
    </div>
  </main>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;

exports[`EdgePage with Mock API renders the edge page title 1`] = `
<div>
  <main
    class="ant-layout-content pageContainer css-dev-only-do-not-override-98ntnt"
    id="edges"
  >
    <div
      class="pageTitleContainer ant-flex css-dev-only-do-not-override-98ntnt ant-flex-wrap-wrap ant-flex-align-center ant-flex-justify-space-between ant-flex-gap-small"
    >
      <span
        class="ant-typography css-dev-only-do-not-override-98ntnt"
        data-testid="heading"
        style="font-weight: 600; font-size: 16px;"
      >
        <div
          class="ant-flex css-dev-only-do-not-override-98ntnt ant-flex-align-center"
          style="gap: 10px;"
        >
          <span
            class="ant-typography css-dev-only-do-not-override-98ntnt"
          >
            Edges
          </span>
          <span
            aria-label="search"
            class="anticon anticon-search cursor-pointer"
            role="img"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              data-icon="search"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
              />
            </svg>
          </span>
        </div>
      </span>
      <button
        class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-link"
        style="font-weight: 500;"
        type="button"
      >
        <span
          class="ant-btn-icon"
        >
          <span
            aria-label="plus-square"
            class="anticon anticon-plus-square"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="plus-square"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
              />
              <path
                d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
              />
            </svg>
          </span>
        </span>
        <span>
          Create Edge
        </span>
      </button>
    </div>
    <div
      class="pageContent"
    >
      <div
        class="mb-sm ant-flex css-dev-only-do-not-override-98ntnt ant-flex-align-center ant-flex-justify-space-between"
        style="margin-top: -10px; gap: 20px;"
      />
      <div
        class="ant-table-wrapper css-dev-only-do-not-override-98ntnt"
      >
        <div
          class="ant-spin-nested-loading css-dev-only-do-not-override-98ntnt"
        >
          <div>
            <div
              aria-busy="true"
              aria-live="polite"
              class="ant-spin ant-spin-spinning css-dev-only-do-not-override-98ntnt"
            >
              <span
                class="ant-spin-dot-holder"
              >
                <span
                  class="ant-spin-dot ant-spin-dot-spin"
                >
                  <i
                    class="ant-spin-dot-item"
                  />
                  <i
                    class="ant-spin-dot-item"
                  />
                  <i
                    class="ant-spin-dot-item"
                  />
                  <i
                    class="ant-spin-dot-item"
                  />
                </span>
              </span>
            </div>
          </div>
          <div
            class="ant-spin-container ant-spin-blur"
          >
            <div
              class="ant-table ant-table-empty css-dev-only-do-not-override-98ntnt"
            >
              <div
                class="ant-table-container"
              >
                <div
                  class="ant-table-content"
                >
                  <table
                    style="table-layout: auto;"
                  >
                    <colgroup>
                      <col
                        style="width: 250px;"
                      />
                      <col
                        style="width: 150px;"
                      />
                      <col
                        style="width: 150px;"
                      />
                      <col
                        style="width: 150px;"
                      />
                    </colgroup>
                    <thead
                      class="ant-table-thead"
                    >
                      <tr>
                        <th
                          aria-label="Name"
                          class="ant-table-cell ant-table-column-has-sorters react-resizable"
                          scope="col"
                          tabindex="0"
                        >
                          <div
                            class="ant-table-filter-column"
                          >
                            <span
                              class="ant-table-column-title"
                            >
                              <div
                                class="ant-table-column-sorters"
                              >
                                <span
                                  class="ant-table-column-title"
                                >
                                  Name
                                </span>
                                <span
                                  class="ant-table-column-sorter ant-table-column-sorter-full"
                                >
                                  <span
                                    aria-hidden="true"
                                    class="ant-table-column-sorter-inner"
                                  >
                                    <span
                                      aria-label="caret-up"
                                      class="anticon anticon-caret-up ant-table-column-sorter-up"
                                      role="img"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        data-icon="caret-up"
                                        fill="currentColor"
                                        focusable="false"
                                        height="1em"
                                        viewBox="0 0 1024 1024"
                                        width="1em"
                                      >
                                        <path
                                          d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                        />
                                      </svg>
                                    </span>
                                    <span
                                      aria-label="caret-down"
                                      class="anticon anticon-caret-down ant-table-column-sorter-down"
                                      role="img"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        data-icon="caret-down"
                                        fill="currentColor"
                                        focusable="false"
                                        height="1em"
                                        viewBox="0 0 1024 1024"
                                        width="1em"
                                      >
                                        <path
                                          d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                        />
                                      </svg>
                                    </span>
                                  </span>
                                </span>
                              </div>
                            </span>
                            <span
                              class="ant-dropdown-trigger ant-table-filter-trigger"
                              role="button"
                              tabindex="-1"
                            >
                              <span
                                aria-label="filter"
                                class="anticon anticon-filter"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="filter"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M349 838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V642H349v196zm531.1-684H143.9c-24.5 0-39.8 26.7-27.5 48l221.3 376h348.8l221.3-376c12.1-21.3-3.2-48-27.7-48z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </div>
                          <span
                            class="react-resizable-handle"
                          />
                        </th>
                        <th
                          aria-label="Cost"
                          class="ant-table-cell ant-table-column-has-sorters react-resizable"
                          scope="col"
                          tabindex="0"
                        >
                          <div
                            class="ant-table-column-sorters"
                          >
                            <span
                              class="ant-table-column-title"
                            >
                              Cost
                            </span>
                            <span
                              class="ant-table-column-sorter ant-table-column-sorter-full"
                            >
                              <span
                                aria-hidden="true"
                                class="ant-table-column-sorter-inner"
                              >
                                <span
                                  aria-label="caret-up"
                                  class="anticon anticon-caret-up ant-table-column-sorter-up"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    data-icon="caret-up"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                    />
                                  </svg>
                                </span>
                                <span
                                  aria-label="caret-down"
                                  class="anticon anticon-caret-down ant-table-column-sorter-down"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    data-icon="caret-down"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                    />
                                  </svg>
                                </span>
                              </span>
                            </span>
                          </div>
                          <span
                            class="react-resizable-handle"
                          />
                        </th>
                        <th
                          class="ant-table-cell react-resizable"
                          scope="col"
                        >
                          <div
                            class="ant-table-filter-column"
                          >
                            <span
                              class="ant-table-column-title"
                            >
                              Labels
                            </span>
                            <span
                              class="ant-dropdown-trigger ant-table-filter-trigger"
                              role="button"
                              tabindex="-1"
                            >
                              <span
                                aria-label="filter"
                                class="anticon anticon-filter"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="filter"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M349 838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V642H349v196zm531.1-684H143.9c-24.5 0-39.8 26.7-27.5 48l221.3 376h348.8l221.3-376c12.1-21.3-3.2-48-27.7-48z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </div>
                          <span
                            class="react-resizable-handle"
                          />
                        </th>
                        <th
                          class="ant-table-cell react-resizable"
                          scope="col"
                        >
                          <div
                            class="ant-table-filter-column"
                          >
                            <span
                              class="ant-table-column-title"
                            >
                              Tags
                            </span>
                            <span
                              class="ant-dropdown-trigger ant-table-filter-trigger"
                              role="button"
                              tabindex="-1"
                            >
                              <span
                                aria-label="filter"
                                class="anticon anticon-filter"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="filter"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M349 838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V642H349v196zm531.1-684H143.9c-24.5 0-39.8 26.7-27.5 48l221.3 376h348.8l221.3-376c12.1-21.3-3.2-48-27.7-48z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </div>
                          <span
                            class="react-resizable-handle"
                          />
                        </th>
                        <th
                          class="ant-table-cell"
                          scope="col"
                        >
                          Actions
                        </th>
                      </tr>
                    </thead>
                    <tbody
                      class="ant-table-tbody"
                    >
                      <tr
                        class="ant-table-placeholder"
                      >
                        <td
                          class="ant-table-cell"
                          colspan="5"
                        >
                          <div
                            class="css-dev-only-do-not-override-98ntnt ant-empty ant-empty-normal"
                          >
                            <div
                              class="ant-empty-image"
                            >
                              <svg
                                height="41"
                                viewBox="0 0 64 41"
                                width="64"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <title>
                                  Simple Empty
                                </title>
                                <g
                                  fill="none"
                                  fill-rule="evenodd"
                                  transform="translate(0 1)"
                                >
                                  <ellipse
                                    cx="32"
                                    cy="33"
                                    fill="#f5f5f5"
                                    rx="32"
                                    ry="7"
                                  />
                                  <g
                                    fill-rule="nonzero"
                                    stroke="#d9d9d9"
                                  >
                                    <path
                                      d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                    />
                                    <path
                                      d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                      fill="#fafafa"
                                    />
                                  </g>
                                </g>
                              </svg>
                            </div>
                            <div
                              class="ant-empty-description"
                            >
                              No data
                            </div>
                          </div>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </main>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;

exports[`EdgePage with Mock API updates an edge: final state after update simulation 1`] = `
<div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;

exports[`EdgePage with Mock API updates an edge: form updated with new values 1`] = `
<div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;

exports[`EdgePage with Mock API updates an edge: initial modal state for update test 1`] = `
<div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;
